network/wsNetwork.go

```diff
@@ -2350,18 +2348,35 @@ func (wn *WebsocketNetwork) updateMessagesOfInterestEnc() {
 	wn.messagesOfInterestEnc = MarshallMessageOfInterestMap(wn.messagesOfInterest)
 	wn.messagesOfInterestEncoded = true
 	atomic.AddUint32(&wn.messagesOfInterestGeneration, 1)
-	wn.messagesOfInterestCond.Broadcast()
+	select {
+	case wn.messagesOfInterestNotify <- struct{}{}:
+	default:
+		// it has been notified enough, it will pick up the latest when it actually runs
+	}
 }
 
 func (wn *WebsocketNetwork) postMessagesOfInterestThread() {
 	var peers []*wsPeer
-	wn.messagesOfInterestMu.Lock()
-	defer wn.messagesOfInterestMu.Unlock()
 	for {
-		wn.messagesOfInterestCond.Wait()
-		peers, _ = wn.peerSnapshot(peers)
-		for _, peer := range peers {
-			wn.maybeSendMessagesOfInterest(peer, wn.messagesOfInterestEnc)
+		select {
+		case <-wn.messagesOfInterestRefresh:
+			// if we're not a relay, and not participating, we don't need txn pool
+			wantTXGossip := wn.nodeInfo.IsParticipating()
+			if wantTXGossip && (wn.wantTXGossip != wantTXGossipYes) {
+				wn.RegisterMessageInterest(protocol.TxnTag)
+				atomic.StoreUint32(&wn.wantTXGossip, wantTXGossipYes)
+			} else if !wantTXGossip && (wn.wantTXGossip != wantTXGossipNo) {
+				wn.DeregisterMessageInterest(protocol.TxnTag)
+				atomic.StoreUint32(&wn.wantTXGossip, wantTXGossipNo)
+			}
+		case <-wn.messagesOfInterestNotify:
```

Comment:
the only methods that enqueue to this channel (in updateMessagesOfInterestEnc) are called via {Register,Deregister}MessageInterest directly above, so this thread is notifying itself via this channel to call maybeSendMessagesOfInterest below. 

What if you set `callMaybeSend = true` under each of the Register/Deregister cases above and then checked it below to decide whether you need to call maybeSendMessagesOfInterest? That would make it possible to get rid of this channe.

Reply:
I was able to unwrap one layer of notification and do it all in one pass through the refresh thread.